packages/style-engine/src/index.ts

```diff
@@ -62,3 +65,20 @@ export function getCSSRules(
 
 	return rules;
 }
+
+/**
+ * Returns an array of classnames.
+ *
+ * @param  style Style object.
+ *
+ * @return string[] An array of classnames.
+ */
+export function getClassnames( style: Style ): string[] {
```

Comment:
I think we could probably think about/ship this functionality separately. 

Just adding it here to evaluate the concept.

Reply:
I like the concept! I think I'd lean toward including it in this PR, it's looking like a good direction to me.

Reply:
Who's responsible for providing the styles for these classnames, while I think having a function that return "state" classnames is fine, I might think that these classnames shouldn't be necessary to style things properly and if we can have all the logic in `generate` function it's a bit better.

Reply:
I guess we can't replace classnames with CSS variables yet in `generate` because that would require block deprecations for now. It might be fine to keep these out for now with a caviat that we may change these APIs later (to fully fulfill the style engine's potential :P)

Also, I don't see this function used in the "style" hook, should we replace the custom logic there with this function (like we did with the inline styles)? 

For Global Styles generation though, it seems that we'll have to implement the CSS variable based styles in generate anyway right?

Reply:
> Who's responsible for providing the styles for these classnames, 

In the post editor, the styles linked to the `has-*` classnames are ultimately pulled from global settings/presents (generated by class-wp-theme-json-gutenberg.php).

An inline JS script hydrates the editor with settings. The styles are passed down [from <Editor />](https://github.com/WordPress/gutenberg/blob/9a195fca8e0a23af1605830b4128edb116a4964a/packages/edit-post/src/editor.js#L160) and [transformed/output in <EditorStyles />](https://github.com/WordPress/gutenberg/blob/efaa86b900f8707a1ae8c11ebf591fde808f24aa/packages/block-editor/src/components/editor-styles/index.js#L71).

Then in [hooks/color.js the classnames](https://github.com/WordPress/gutenberg/blob/6de16981c7fef3db7cc4a9d6cfcb0ad85c916c62/packages/block-editor/src/hooks/color.js#L247) for the block wrapper are compiled and added to the block props.

Is that what you meant by "responsible"?

The last step is what I had in mind for the style engine to take over initially, only because we're doing it in the backend. I was playing around with generating the `has-*` classnames in hooks/color.js via the style engine this morning. It works okay, but I get the feeling I'm building "around" existing logic, so not sure yet.

> while I think having a function that return "state" classnames is fine, I might think that these classnames shouldn't be necessary to style things properly and if we can have all the logic in generate function it's a bit better.

Now I'm writing about it, it might be a better approach to have the style engine generate the CSS styles at the [transformed/output stage mentioned above](https://github.com/WordPress/gutenberg/blob/efaa86b900f8707a1ae8c11ebf591fde808f24aa/packages/block-editor/src/components/editor-styles/index.js#L71) and forget about block classnames for now.

> I guess we can't replace classnames with CSS variables yet in generate because that would require block deprecations for now. 

Yeah, I did have a shot at putting the CSS vars inline... and it worked 🎉  but the console exploded in my face 😄 

> Also, I don't see this function used in the "style" hook, should we replace the custom logic there with this function (like we did with the inline styles)?

I tested it out this morning, and have things half-working. There are some inexplicable fixture-related errors, but I'll push anyway so can eyeball the code.

> For Global Styles generation though, it seems that we'll have to implement the CSS variable based styles in generate anyway right?

Yeah for sure. That will be its own challenge. It makes me think that I should leave the block classname alone for now and rather devote the time into centralizing _style_ generation.

Reply:
> It makes me think that I should leave the block classname alone for now and rather devote the time into centralizing style generation.

I've pushed my work so far in https://github.com/WordPress/gutenberg/pull/40665/commits/393fac42c75ad97c5d397255fd4f9aefa9cefc25 so we can look.

Reply:
> Yeah, I did have a shot at putting the CSS vars inline... and it worked 🎉 but the console exploded in my face 😄

But isn't this the right thing to do anyway. It will solve both "block style generation" and "global styles generation" at the same time with the same style engine. 

I agree that we need to find a way to either:

 1- Don't use the style engine yet for colors and such in the hooks/color.js for now and keep the existing logic there
 2- Have some kind of flag per block to say: this block is ready for full style engine style generation which "for me" means:
     - do not serialize inline styles in the saved content
     - rely on the JS style engine for the "edit" function (injected somewhere in hooks/color)
     - rely on the PHP style engine for the frontend.

Maybe it's too soon for 2, I don't know.

---
https://github.com/WordPress/gutenberg/commit/393fac42c75ad97c5d397255fd4f9aefa9cefc25 is fine if we really want to keep getClassnames to centralize the "classname" generation for blocks based on styles, but for me, this is not really the style engine, or more precisely, not the main purpose of the style engine.

I hope that makes a bit of sense :)

Reply:
> Don't use the style engine yet for colors and such in the hooks/color.js for now and keep the existing logic there

Yeah, what this PR is attempting might be too much too soon. There's probably more value in trying to get style/css generation right across blocks/global styles.

> rely on the PHP style engine for the frontend.

It would be fabulous for the PHP to take care of all frontend styles. Not sure about inline CSS saved in the editor though... We could maybe start over at layout.php for now.

> want to keep getClassnames to centralize the "classname" generation for blocks based on styles, but for me, this is not really the style engine, or more precisely, not the main purpose of the style engine.

I get you. My idea was to take as much decision making about classnames and styles away from the editor as possible. Classname generation might be a bit premature I'd admit that. 

I'd be fine with skipping this functionality in the JS in this PR. Maybe even revisiting the backend functionality too.

 I'm still holding onto the hope that it'll all be clearer after we experiment with things. 🤞 Am I crazy? 😄 

> I hope that makes a bit of sense :)

Yes, it's given us a lot to think about, thanks a lot!

Reply:
> Maybe it's too soon for 2, I don't know.

Thanks for laying out a bit of how that could work — I think that sounds like something to pursue in a separate PR (perhaps once we've got all the block supports implemented server-side?) but I really like the idea of a per-block flag to opt-in for the style engine to render out the styles server-side on the site frontend. We could also use that flag to conditionally ignore style validation errors as a gentle way of gradually introducing the work from https://github.com/WordPress/gutenberg/pull/37942 (your PR to ignore style validation errors) so we can avoid issues with deprecations.

Reply:
Thanks for clarifying 👍 I'm confident this is going in the right direction ❤️